### ISSUE TITLE
About the Few Shot experiment

### ISSUE THREAD

**Issue by qmechlab** [TreeLLi/APT#8]
Hi, thanks for your excellent work. I am using this command (bash scripts/APT.sh dtd vit_b32_ep200 end 16 16 False 4 2.67 3 0 onfly 0) to fine-tune prompts on the dtd dataset. I have the following questions.

1. I got an accuracy of 45.7%. But the accuracy in the training log you provided is 21.8%. I noticed that you commented out a sentence in the test() function in the APT class (# input = attack(input, label).detach()). So I'm guessing the current version of the code outputs clean accuracy and the log you provided outputs adversarial accuracy. Is this the correct understanding?

2. If the understanding of question 1 is correct, then can I directly test the clean accuracy and adversarial accuracy in the test function in the APT class? Should the results obtained be the same as those obtained by running the python evaluate.py path_to_checkpoint --cls-prompt prompter --attack pgd command? I made this attempt and the results are close, but not exactly the same. I just want to make sure that if I only care about the pgd attack, can I just use the results given by the APT.test() function without having to run evaluate.py?

3. I notice that you provide “perturbed” attacks for training, but the four attacks for evaluation do not have "perturbed" mode. So I guess "perturbed" mode is only used for adversarial training? Is it possible to use the attack of "perturbed"  mode at evaluation time as well?

Thank you for all the help you can provide.

**Comment by TreeLLi** [comment]
Hi,

thanks much for your interest in our work! Please find below my response to your questions:

1. You are right. The current version does not run adversarial evaluation in the `test()` function. I may comment that line in some experiments to catch up the deadline. This should not affect the results since we did not use the results from `test()` directly nor to select the checkpoint to report.
2. I think you can use `test()` to get both clean and adv accuracy. The reason why two results may not be exactly the same is that `test()` used 10 steps of PGD while `evaluate.py` used 100 steps. However, I have not tested if they really give the same results. I will be happy to know if you have some results to share.
3. Yes, `perturbed` mode was initially experimented only in adversarial training. It can be used in evaluation. This is actually an interesting direction to explore. If you check the P4 row in Figure 4 in the paper, we noticed that using an attack prompt different from the inference prompt can improve the effectiveness of attack. This suggests that "perturbing" (optimizing) text prompts while generating visual adversarial examples may produce stronger visual adversarial examples.

Please let me know if anything else I could help.

All best,
Lin

**Comment by qmechlab** [comment]
Hi Lin,

Thank you for your detailed answer. I have conducted experiments on the dtd dataset. I added the experimental result of PGD100 on APT.test(). The records are as follows:

```
	       clean	PGD10	PGD100
APT.test()      45.7	21.2	21.1
evaluate.py	45.8	none	20.69
```

The difference in experimental results between evaluate and test on Clean and PGD100 metrics could be the effect of random numbers?


Also, notice that you used four datasets ImageNet-V2 [52], ImageNet-Sketch [62], ImageNet-R [23] and ObjectNet [4] in your OOD experiment. For this I encountered three problems:
1. I get an error when evaluating ImageNet-Sketch and ImageNet-V2 using evaluate.py. I noticed that you seem to have only written the ports for the 'ImageNetR', 'ImageNetA', 'ON' datasets in evaluate.py. There is no mention of the ImageNet-Sketch and ImageNet-V2 datasets.

2. I noticed that the CoOP method includes the processing of ImageNetR, ImageNet-Sketch and ImageNet-V2 datasets. Can I use its processing code and then evaluate these datasets directly in a zeroshot manner? Should this be the result of OOD?

3. Is it proper to use the ImageNetA dataset in OOD experiments? Noticed that you have implemented it in your code, but there is no mention of this dataset in the paper.

Many thanks for your help.

**Comment by TreeLLi** [comment]
Hi,

sorry for coming back late and thanks for sharing these results! I guess random numbers could be a reason, however, we did not use the result from `.test()` so I did not go in depth the implementation of this function (it was inherited from CoOp). Have u tried multiple runs and compared the average?

To answer your questions about OOD experiments:

1. We used the default loading pipeline from CoOp to load these two datasets. You should be able to evaluate them by adding the argument `--dataset ImageNetSketch/ImageNetV2`.
2. Yes, we also used the CoOp pipeline for ImageNetV2 and ImageNetSketch, while we used our own implementation of ImageNetR. Nevertheless, they should not affect the results.
3. ImageNetA is a setting of OOD. The reason we did not use is that ImageNetA has a overly strong distribution shift leading to very low clean accuracy and close to 0s robust accuracy for all evaluated baselines in our preliminary tests, which makes it hard to compare the performance of different methods.

Best wishes,
Lin

**Comment by qmechlab** [comment]
Hi Lin,

Thank you for your reply.  I added “import datasets.imagenetv2” to evaluate.py. Then I configured the dataset using CoOp's instructions DATASETS.md. This is different from your DATASETS.md (for imagenet-v2). I finally got it to work, but I wonder if this is the right configuration?

Both the zero-shot experiment and the OOD experiment use weights pre-trained on imagenet. However, in this paper, the weights are trained with different hyperparameters (e.g. M and numshots, etc.). Why do we use different hyperparameters to train the weights? Is it ok to use the trained weights (zs_imagenet_vitb32_eps4_nctx1_100shots_UC) used for zero-shot for OOD experiments?

Many thanks for your help.

**Comment by TreeLLi** [comment]
Hi,

yes, they should use the same data source. You could also do a sanity check to see if your result is close to our reported numbers.

Different hyperparameters were used to optimize the performance. Yes, you can because it is trained on the training set of ImageNet and therefore have not seen any (training) data of those OOD datasets.

**Comment by qmechlab** [comment]
Thank you for your prompt reply. Hope you are successful in your studies!
